### PR TITLE
Paradigm Restructuring Part 4 - Process

### PR DESCRIPTION
**Pull request by nupplaphil**
FollowUp #10921 

This one moves all "process" aspects into a own package "Worker".
The process entity is now passed into the `Worker`-methods itself rather than created them inside. From my point of view, it's now better understandable when we create a `Process` and how/when we delete it.

ToDo:
- [x] Inject the new process id into the Logger itself after forking :)

### CODE REVIEW

**Comment by nupplaphil** [comment]
@annando / @MrPetovan - tbh .. what do we do with the `process` table at all?
It was introduced here https://github.com/friendica/friendica/pull/2802 but never used except creating/deleting it.

Is it a leftover from the past? If so I'd remove the whole table instead of refactoring it ..

[edit] OK I found one occurrence here: https://github.com/friendica/friendica/blob/c9adf4ca1d0adc2ac83c2bd6aeec5caa599f5cab/src/Core/Worker.php#L809

**Comment by nupplaphil** [comment]
Seems to work.

Log:
```console
2021-10-31T13:43:13Z worker [NOTICE]: User 1 posted home timeline item 6094899 {"worker_id":"b99fc42","worker_cmd":"twitter_sync_run"} - {"file":"twitter.php","line":1993,"function":"twitter_fetchhometimeline","uid":"fb3362","process_id":45}                                          
2021-10-31T13:43:14Z daemon [NOTICE]: Load: 2.36279296875/20 - processes: 7/0/46 (0:0, 40:8) - maximum: 7/10 [] - {"file":"Worker.php","line":777,"function":"tooMuchWorkers","uid":"2b0691","process_id":26}                                                                              
2021-10-31T13:43:14Z daemon [NOTICE]: Load: 2.36279296875/20 - processes: 7/0/44 (0:0, 30:1, 40:8) - maximum: 7/10 [] - {"file":"Worker.php","line":777,"function":"tooMuchWorkers","uid":"2b0691","process_id":26}                                                                        
2021-10-31T13:43:14Z worker [NOTICE]: Load: 2.36279296875/20 - processes: 7/10/43 (0:1, 40:9) - maximum: 7/10 [] - {"file":"Worker.php","line":781,"function":"tooMuchWorkers","uid":"ca280d","process_id":248}                                                                            
2021-10-31T13:43:14Z worker [NOTICE]: Active worker limit reached, quitting. [] - {"file":"Worker.php","line":182,"function":"isReady","uid":"ca280d","process_id":248}                                                                                                                    
2021-10-31T13:43:14Z daemon [NOTICE]: Load: 2.36279296875/20 - processes: 7/0/43 (0:0, 30:1, 40:8) - maximum: 7/10 [] - {"file":"Worker.php","line":777,"function":"tooMuchWorkers","uid":"2b0691","process_id":26}                                                                        
2021-10-31T13:43:14Z worker [NOTICE]: Load: 2.36279296875/20 - processes: 7/10/42 (0:1, 40:9) - maximum: 7/10 [] - {"file":"Worker.php","line":781,"function":"tooMuchWorkers","uid":"f43bcc","process_id":251}                                                                            
2021-10-31T13:43:14Z worker [NOTICE]: Active worker limit reached, quitting. [] - {"file":"Worker.php","line":182,"function":"isReady","uid":"f43bcc","process_id":251}                                                                                                                    
2021-10-31T13:43:14Z worker [NOTICE]: Load: 2.36279296875/20 - processes: 7/10/42 (0:1, 40:9) - maximum: 7/10 [] - {"file":"Worker.php","line":781,"function":"tooMuchWorkers","uid":"af6c43","process_id":254}                                                                            
2021-10-31T13:43:14Z worker [NOTICE]: Active worker limit reached, quitting. [] - {"file":"Worker.php","line":182,"function":"isReady","uid":"af6c43","process_id":254}                                                                                                                    
2021-10-31T13:43:14Z daemon [NOTICE]: Load: 2.36279296875/20 - processes: 7/0/42 (0:0, 30:1, 40:8) - maximum: 7/10 [] - {"file":"Worker.php","line":777,"function":"tooMuchWorkers","uid":"2b0691","process_id":26}                                                                        
2021-10-31T13:43:14Z worker [NOTICE]: Load: 2.36279296875/20 - processes: 7/10/41 (0:1, 40:9) - maximum: 7/10 [] - {"file":"Worker.php","line":781,"function":"tooMuchWorkers","uid":"766293","process_id":256}                                                                            
2021-10-31T13:43:14Z worker [NOTICE]: Active worker limit reached, quitting. [] - {"file":"Worker.php","line":182,"function":"isReady","uid":"766293","process_id":256}                                                                                                                    
2021-10-31T13:43:14Z daemon [NOTICE]: Load: 2.36279296875/20 - processes: 7/0/42 (0:0, 40:8) - maximum: 7/10 [] - {"file":"Worker.php","line":777,"function":"tooMuchWorkers","uid":"2b0691","process_id":26}                                                                              
2021-10-31T13:43:14Z worker [NOTICE]: Load: 2.36279296875/20 - processes: 7/10/40 (0:1, 40:9) - maximum: 7/10 [] - {"file":"Worker.php","line":781,"function":"tooMuchWorkers","uid":"dad6e0","process_id":259}                                                                            
2021-10-31T13:43:14Z worker [NOTICE]: Active worker limit reached, quitting. [] - {"file":"Worker.php","line":182,"function":"isReady","uid":"dad6e0","process_id":259}
```

Shortly after:

```mysql
MariaDB [friendica]> select * from process;
+-----+------------+---------------------+
| pid | command    | created             |
+-----+------------+---------------------+
|  37 | worker.php | 2021-10-31 13:42:52 |
|  39 | worker.php | 2021-10-31 13:42:52 |
|  41 | worker.php | 2021-10-31 13:42:52 |
|  43 | worker.php | 2021-10-31 13:42:52 |
|  45 | worker.php | 2021-10-31 13:42:52 |
|  47 | worker.php | 2021-10-31 13:42:52 |
|  51 | worker.php | 2021-10-31 13:42:52 |
|  54 | worker.php | 2021-10-31 13:42:52 |
|  82 | worker.php | 2021-10-31 13:42:53 |
+-----+------------+---------------------+
9 rows in set (0.001 sec)

MariaDB [friendica]> select * from process;
+-----+------------+---------------------+
| pid | command    | created             |
+-----+------------+---------------------+
|  37 | worker.php | 2021-10-31 13:42:52 |
|  39 | worker.php | 2021-10-31 13:42:52 |
|  41 | worker.php | 2021-10-31 13:42:52 |
|  43 | worker.php | 2021-10-31 13:42:52 |
|  47 | worker.php | 2021-10-31 13:42:52 |
|  51 | worker.php | 2021-10-31 13:42:52 |
|  54 | worker.php | 2021-10-31 13:42:52 |
|  82 | worker.php | 2021-10-31 13:42:53 |
+-----+------------+---------------------+
8 rows in set (0.000 sec)

MariaDB [friendica]> select * from process;
+-----+------------+---------------------+
| pid | command    | created             |
+-----+------------+---------------------+
|  41 | worker.php | 2021-10-31 13:42:52 |
+-----+------------+---------------------+
1 row in set (0.000 sec)

MariaDB [friendica]> select * from process;
Empty set (0.001 sec)

```

**Comment by nupplaphil** [comment]
aaand done:
```console
2021-10-31T19:26:03Z daemon [INFO]: Spawned new worker {"pid":23} - {"file":"Worker.php","line":1133,"function":"forkProcess","uid":"75ae60","process_id":22}                                                                                                                               
2021-10-31T19:26:03Z daemon [INFO]: Worker spawned {"pid":23,"wait_cycles":0} - {"file":"Worker.php","line":1155,"function":"forkProcess","uid":"afdfe6","process_id":23}              
```

First line is the parent, second line is the forked worker :-)

File: src/App.php
Comment (MrPetovan) [inline src/App.php]: Formatting
```suggestion
	 * @param Core\System                 $system   The system methods
```

File: src/Core/System.php
Comment (MrPetovan) [inline src/Core/System.php:107]: On Linux systems, this check doesn't make much sense since Linux will use as much memory as possible for internal caching, even when performance is nominal.

I think it should be deactivated like the `isMaxProcessesReached` test. Friendica has become a fully-fledged web application that requires external monitoring rather than these hacks.

File: src/Core/System.php
Comment (MrPetovan) [inline src/Core/System.php]: We should remove this case, it doesn't makes sense to bar access to the frontend, Friendica is more demanding resource-wise than before, so we shouldn't try to accommodate anemic servers by default. Worker load limit makes more sense because the background tasks can be delayed a bit without being very noticeable like a blank page for the frontend.

File: src/Core/System.php
Comment (MrPetovan) [inline src/Core/System.php]: Formatting
```suggestion
		}

		if (!is_resource($resource)) {
```

File: src/Core/System.php
Comment (MrPetovan) [inline src/Core/System.php]: Formatting
```suggestion
		}

		proc_close($resource);
```

File: src/Core/Worker.php
Comment (MrPetovan) [inline src/Core/Worker.php]: The added use statement enables using the shorthand
```suggestion
	/** @var Process */
```

File: src/Core/Worker/Repository/Process.php
Comment (MrPetovan) [inline src/Core/Worker/Repository/Process.php]: Shouldn't this be a method parameter where we can pass `basename(__FILE__)` in the call instead?

File: src/Core/Worker/Repository/Process.php
Comment (MrPetovan) [inline src/Core/Worker/Repository/Process.php:116]: Shouldn't we try to rollback here?

File: src/DI.php
Comment (MrPetovan) [inline src/DI.php:242]: Why do we need to re-add the same Logger rules?

File: src/DI.php
Comment (MrPetovan) [inline src/DI.php:241]: Why do we need to re-add the same Logger rules?

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: src/DI.php
Comment (nupplaphil) [reply src/DI.php:242]: Because this wipes the private instance cache of Dice for these classes, so inside a child, the logger will get build again

File: src/Core/System.php
Comment (nupplaphil) [reply src/Core/System.php]: I removed all the `maxloadavg_frontend` usages, see https://github.com/friendica/friendica/pull/10941/commits/8179af3701209ba9d674be2929edaa0884ae438d

File: src/Core/System.php
Comment (nupplaphil) [reply src/Core/System.php:107]: Done, see https://github.com/friendica/friendica/pull/10941/commits/b61532f77b02128a22ef9f850c6793b008424cb6

**Comment by nupplaphil** [comment]
@annando / @tobiasd / @alfredsk
We're deactivating some general memory-restrictions:
- No more Frontend restrictions: https://github.com/friendica/friendica/pull/10941/commits/8179af3701209ba9d674be2929edaa0884ae438d
- Deactivating Min-Memory check: https://github.com/friendica/friendica/pull/10941/commits/8179af3701209ba9d674be2929edaa0884ae438d

Are there any doubts? just 2 be sure :-)

**Review by MrPetovan — APPROVED** [review]
Waiting on other people's approval given the sensitivity of the changes.

**Comment by annando** [comment]
@nupplaphil When I read the code correctly the change in the frontend load check part also removed the check if the system is running in the frontend or backend. So the limit check is now done equally for both frontend and backend. Is this wanted?

**Comment by nupplaphil** [comment]
Oups, thx for the hint!

**Comment by nupplaphil** [comment]
hmm checked again .. 

I removed the check inside `App::runFrontend()`, which is only called by `index.php`, so it's the foreground process. The background process is checked here: 
https://github.com/friendica/friendica/blob/e7275272f5f9c49ccfa0a78f82934be0373f831f/bin/daemon.php#L199
and here:
https://github.com/friendica/friendica/blob/65e889b26385b85f8d6e2735a8d6d4263cdd0c80/src/Core/Worker.php#L71

**Comment by annando** [comment]
@nupplaphil have you checked how the logs are looking inside the worker processes? It would be great if each individual worker task execution had a unique id so that all operations in the log could be grouped on a single task execution.

**Comment by nupplaphil** [comment]
Definitions for the logger context inside the worker
- `process_id` = PID of the current working process (the whole `Worker.php`)
- `worker_id` = Unique ID for each workerqueue task execution
- `worker_cmd` = The current Worker command

@annando - Did you mean it that way? :)

**Comment by annando** [comment]
Currently this `worker_id` is not unique:
```
2021-11-05T02:52:24Z worker [INFO]: Process start. {"priority":20,"id":204686790,"worker_id":"3f5e356","worker_cmd":"SpoolPost"} - {"file":"Worker.php","line":471,"function":"execFunction","uid":"6616a0","process_id":9460}
2021-11-05T02:52:24Z worker [INFO]: Process start. {"priority":30,"id":204686791,"worker_id":"3f5e356","worker_cmd":"Cron"} - {"file":"Worker.php","line":471,"function":"execFunction","uid":"6616a0","process_id":9460}
2021-11-05T02:52:25Z worker [INFO]: Process start. {"priority":30,"id":204686793,"worker_id":"3f5e356","worker_cmd":"PollContacts"} - {"file":"Worker.php","line":471,"function":"execFunction","uid":"6616a0","process_id":9460}
2021-11-05T02:52:25Z worker [INFO]: Process start. {"priority":30,"id":204686800,"worker_id":"3f5e356","worker_cmd":"twitter_sync_run"} - {"file":"Worker.php","line":471,"function":"execFunction","uid":"6616a0","process_id":9460}
2021-11-05T02:52:25Z worker [INFO]: Process start. {"priority":30,"id":204686808,"worker_id":"3f5e356","worker_cmd":"OnePoll"} - {"file":"Worker.php","line":471,"function":"execFunction","uid":"6616a0","process_id":9460}
2021-11-05T02:52:26Z worker [INFO]: Process start. {"priority":30,"id":204686827,"worker_id":"3f5e356","worker_cmd":"twitter_sync_run"} - {"file":"Worker.php","line":471,"function":"execFunction","uid":"6616a0","process_id":9460}
2021-11-05T02:52:26Z worker [INFO]: Process start. {"priority":40,"id":204686839,"worker_id":"3f5e356","worker_cmd":"UpdateServerDirectory"} - {"file":"Worker.php","line":471,"function":"execFunction","uid":"6616a0","process_id":9460}
2021-11-05T02:52:26Z worker [INFO]: Process start. {"priority":40,"id":204686848,"worker_id":"3f5e356","worker_cmd":"UpdateContact"} - {"file":"Worker.php","line":471,"function":"execFunction","uid":"6616a0","process_id":9460}
```

**Comment by nupplaphil** [comment]
with https://github.com/friendica/friendica/pull/10941/commits/9a5e6642b64f9fafe0afbbcf061cf626f57be3a4 it is now:
```console
2021-11-05T20:46:05Z worker [INFO]: Process start. {"priority":20,"id":24686958,"worker_id":"b28bf4c","worker_cmd":"SpoolPost"} - {"file":"Worker.php","line":475,"function":"execFunction","uid":"e2076f","process_id":403}                                                                
2021-11-05T20:46:05Z worker [INFO]: Process start. {"priority":30,"id":24686959,"worker_id":"3d29d64","worker_cmd":"Cron"} - {"file":"Worker.php","line":475,"function":"execFunction","uid":"e2076f","process_id":403}                                                                     
2021-11-05T20:46:05Z worker [INFO]: Process start. {"priority":30,"id":24686960,"worker_id":"09e8309","worker_cmd":"ForkHook"} - {"file":"Worker.php","line":475,"function":"execFunction","uid":"e2076f","process_id":403}                                                                
2021-11-05T20:46:05Z worker [INFO]: Process start. {"priority":30,"id":24686961,"worker_id":"464d2fb","worker_cmd":"ForkHook"} - {"file":"Worker.php","line":475,"function":"execFunction","uid":"e2076f","process_id":403}                                                                
2021-11-05T20:46:05Z worker [INFO]: Process start. {"priority":30,"id":24686962,"worker_id":"cb8214e","worker_cmd":"PollContacts"} - {"file":"Worker.php","line":475,"function":"execFunction","uid":"e2076f","process_id":403}                                                            
2021-11-05T20:46:05Z worker [INFO]: Process start. {"priority":30,"id":24686966,"worker_id":"027c2a3","worker_cmd":"ExpireAndRemoveUsers"} - {"file":"Worker.php","line":475,"function":"execFunction","uid":"e2076f","process_id":403}                                                    
2021-11-05T20:46:05Z worker [INFO]: Process start. {"priority":30,"id":24686968,"worker_id":"8cbd31a","worker_cmd":"twitter_sync_run"} - {"file":"Worker.php","line":475,"function":"execFunction","uid":"e2076f","process_id":403}                                                        
2021-11-05T20:46:05Z worker [INFO]: Process start. {"priority":40,"id":24686963,"worker_id":"e583206","worker_cmd":"UpdateContacts"} - {"file":"Worker.php","line":475,"function":"execFunction","uid":"0ed032","process_id":407}                                                          
2021-11-05T20:46:05Z worker [INFO]: Process start. {"priority":30,"id":24686969,"worker_id":"99ab5fd","worker_cmd":"twitter_sync_run"} - {"file":"Worker.php","line":475,"function":"execFunction","uid":"25b81f","process_id":409}                                                        
2021-11-05T20:46:05Z worker [INFO]: Process start. {"priority":40,"id":24686964,"worker_id":"c9423ee","worker_cmd":"UpdateGServers"} - {"file":"Worker.php","line":475,"function":"execFunction","uid":"0a78a7","process_id":411}                                                          
2021-11-05T20:46:05Z worker [INFO]: Process start. {"priority":40,"id":24686965,"worker_id":"a7d8386","worker_cmd":"UpdateServerDirectories"} - {"file":"Worker.php","line":475,"function":"execFunction","uid":"aedc33","process_id":414}                                                 
2021-11-05T20:46:05Z worker [INFO]: Process start. {"priority":40,"id":24686967,"worker_id":"bc0038c","worker_cmd":"PostUpdate"} - {"file":"Worker.php","line":475,"function":"execFunction","uid":"0a78a7","process_id":411}                                                              
2021-11-05T20:46:05Z worker [INFO]: Process start. {"priority":40,"id":24686970,"worker_id":"a03e50d","worker_cmd":"UpdateServerPeers"} - {"file":"Worker.php","line":475,"function":"execFunction","uid":"9eb596","process_id":421}                                                 
```

**Review by annando — APPROVED** [review]

**Comment by MrPetovan** [comment]
Fixes #10871

**Comment by bkil** [comment]
Related to #10131

**Pull request merged**
